Huffman compression fixed thanks to gba-huff

GValiente opened the pull request:
Huffman compression is replaced with [gba-huff](https://github.com/GValiente/gba-huff) compressor.

It fixes https://github.com/devkitPro/grit/issues/4, and I haven't had any issues with other data I have tested.

Since gba-huff is licensed under the zlib license, there should be no issues integrating it with grit.

mtheall commented:
Unfortunately still broken. Tested using nds 256-color bitmap background example but changed to huff compression.

Test image:
![gradient](https://user-images.githubusercontent.com/896483/179544676-911f4c2c-c8f9-49ff-837e-1c450a6bc187.png)

Original grit:
![image](https://user-images.githubusercontent.com/896483/179544476-11c25b02-c016-477a-af57-259b07663f46.png)

Pull request grit:
![image](https://user-images.githubusercontent.com/896483/179544752-28e1ed2e-4d08-45c4-b6e6-0e8b572822ae.png)

GValiente commented:
Thanks for looking at it.

What do I have to change in [this example](https://github.com/devkitPro/nds-examples/tree/master/Graphics/Backgrounds/256_color_bmp) to reproduce your test? (load code, grit flags, etc)

mtheall commented:
The important part is that the input data is uniformly distributed. The huffman tree encoding breaks when one of the subtrees has more than 63 child nodes.

data/drunkenlogo.grit:
add `-gzh`

data/drunkenlogo.png:
replace with grayscale gradient image

source/main.cpp:
replace `dmaCopy(drunkenlogoBitmap, bgGetGfxPtr(bg3), 256*256);` with `decompress(drunkenlogoBitmap, bgGetGfxPtr(bg3), HUFF);`

GValiente commented:
I'm not able to reproduce it with MelonDS (with and without external BIOS files) and with DeSmuME.

I get a black screen instead (it works fine when no compression is used).

I have also tried with RLE compression (which works fine with grit) and I get the same black screen.

To try to test if RLE works:
* I have added `-gzr` to data/drunkenlogo.grit
* I have replaced `dmaCopy(drunkenlogoBitmap, bgGetGfxPtr(bg3), 256*256);` with `decompress(drunkenlogoBitmap, bgGetGfxPtr(bg3), RLE);` in source/main.cpp

At last, I have tried to replace the image used in #4 with a 64x64 4bpp image like the one you have posted and again it works fine.

mtheall commented:
Does this [branch](https://github.com/devkitPro/grit/tree/huffman) work for you?

GValiente commented:
> Does [this branch](https://github.com/devkitPro/grit/tree/huffman) work for you?

No, it breaks #4.

mtheall commented:
Pushed up a [new commit](https://github.com/devkitPro/grit/commit/20d951e86e3798c814f19dd19606ffc29a1e4ded).
Appears to fix #4 for me. Seems that GBA requires the tree to occupy 0x200 bytes? NDS doesn't, so that'd be wasted space there.

GValiente commented:
Now it works great with everything I have tried :)

GValiente commented:
> Pushed up a [new commit](https://github.com/devkitPro/grit/commit/20d951e86e3798c814f19dd19606ffc29a1e4ded). Appears to fix #4 for me. Seems that GBA requires the tree to occupy 0x200 bytes? NDS doesn't, so that'd be wasted space there.

Are you going to merge your branch?

GValiente commented:
I'll close this pull request then :)

Thanks a lot for your help.

tom42 commented:
> Pushed up a [new commit](https://github.com/devkitPro/grit/commit/20d951e86e3798c814f19dd19606ffc29a1e4ded). Appears to fix #4 for me. Seems that GBA requires the tree to occupy 0x200 bytes? NDS doesn't, so that'd be wasted space there.

Are you sure this is true? I have a test program where I decode a file to VRAM. It's using 8 bit huffman encoding. The tree size byte is 0x07, so that would be 16 bytes. The file has been encoded using CUE's encoder from https://github.com/PeterLemon/Nintendo_DS_Compressors

I've tested it on a GBA SP, where it works.

What I do _think_ is required is that the tree size is a multiple of 4 bytes.

GBATEK states that the

> Compressed Bitstream [is] (stored in units of 32bits).

Presumably this is so that the bit stream can be read using word accesses on an ARM CPU.

GBATEK further states that the tree size byte is the

> Size of Tree Table/2-1 (ie. Offset to Compressed Bitstream)

Note that size and offset is not really the same. I think there need to be padding bytes at the end of the tree table (that's why he calls it offset), so that the bitstream ends up at a word aligned address, but that's all.

I can supply that test program if interested.
